In [1]:
!pip -q install trl peft bitsandbytes>=0.46.1

In [2]:
import torch
from collections import Counter
from datasets import load_dataset

In [3]:
from huggingface_hub import login

login(token="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")

## Load model

In [4]:
from transformers import AutoModelForCausalLM, AutoTokenizer

HF_MODEL_ID = "quangne/text2diagram-AceMath-1.5B-Instruct-merged-1k"

model = AutoModelForCausalLM.from_pretrained(
    HF_MODEL_ID,
    trust_remote_code=True,
    device_map="auto",
)

tokenizer = AutoTokenizer.from_pretrained(HF_MODEL_ID, trust_remote_code=True, use_fast=True)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/3.09G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/338 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/117 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/605 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/616 [00:00<?, ?B/s]

chat_template.jinja:   0%|          | 0.00/435 [00:00<?, ?B/s]

## Evaluation on Test Split

This section computes:
- Exact Match (EM)
- Valid DSL Rate (simple syntax checks)
- Fact-level Precision/Recall/F1 (line-based DSL facts)

Run the next cell after loading `model` and `tokenizer`.

In [5]:
DATASET_WORKSPACE = "quangne"
DATASET_REPO = "geometry"

INSTRUCTION_TEMPLATE: str = (
   "Chuyển bài toán hình học tiếng Việt sang Geometry DSL (S-expression).\n"
   "Chỉ trả về DSL thuần văn bản hợp lệ từ đề bài, không markdown, không giải thích.\n"
   "Bỏ qua phần yêu cầu chứng minh hoặc câu hỏi phụ, nhưng giữ mọi dữ kiện hình học và điều kiện ràng buộc trong đề.\n\n"
   "Đề bài:\n{query}\n\n"
   "DSL:"
)

In [6]:
from collections import Counter
import time
from datasets import load_dataset

try:
    from tqdm.auto import tqdm
except ImportError:
    tqdm = None

if torch.cuda.is_available():
    torch.backends.cuda.matmul.allow_tf32 = True
    torch.backends.cudnn.allow_tf32 = True


def _normalize_text(s: str) -> str:
    if s is None:
        return ""
    lines = [" ".join(line.strip().split()) for line in str(s).strip().splitlines() if line.strip()]
    return "\n".join(lines)


def _extract_facts(dsl: str):
    """Line-based facts for a lightweight structural metric."""
    norm = _normalize_text(dsl)
    if not norm:
        return []
    return [line for line in norm.splitlines() if line]


def _is_valid_dsl(dsl: str) -> bool:
    """Basic syntax checks: balanced parentheses + line shape."""
    text = _normalize_text(dsl)
    if not text:
        return False

    balance = 0
    for ch in text:
        if ch == "(":
            balance += 1
        elif ch == ")":
            balance -= 1
            if balance < 0:
                return False
    if balance != 0:
        return False

    for line in text.splitlines():
        if not (line.startswith("(") and line.endswith(")")):
            return False
    return True


def _fact_prf1(pred_facts, ref_facts):
    pred_counter = Counter(pred_facts)
    ref_counter = Counter(ref_facts)
    overlap = sum((pred_counter & ref_counter).values())
    pred_total = sum(pred_counter.values())
    ref_total = sum(ref_counter.values())
    precision = overlap / pred_total if pred_total else 0.0
    recall = overlap / ref_total if ref_total else 0.0
    f1 = (2 * precision * recall / (precision + recall)) if (precision + recall) else 0.0
    return precision, recall, f1


def _prepare_inference_context():
    model.eval()
    use_cuda = torch.cuda.is_available() and str(model.device).startswith("cuda")
    compute_dtype = torch.bfloat16 if (use_cuda and torch.cuda.is_bf16_supported()) else torch.float16

    # Some PEFT/quantized runs keep lm_head in float32 while hidden states are bf16/fp16.
    if hasattr(model, "lm_head") and hasattr(model.lm_head, "to"):
        try:
            model.lm_head = model.lm_head.to(dtype=compute_dtype)
        except Exception:
            pass

    return use_cuda, compute_dtype


def generate_dsl(
    instruction: str,
    max_new_tokens: int = 256,
    use_cuda: bool = False,
    compute_dtype: torch.dtype = torch.float16,
) -> str:
    prompt_body = INSTRUCTION_TEMPLATE.format(query=instruction)
    if tokenizer.chat_template:
        prompt = tokenizer.apply_chat_template(
            [{"role": "user", "content": prompt_body}],
            tokenize=False,
            add_generation_prompt=True,
        )
    else:
        prompt = f"User:\n{prompt_body}\n\nAssistant:\n"

    inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

    with torch.inference_mode():
        if use_cuda:
            with torch.autocast(device_type="cuda", dtype=compute_dtype):
                outputs = model.generate(
                    **inputs,
                    max_new_tokens=max_new_tokens,
                    do_sample=False,
                    repetition_penalty=1.08,
                    eos_token_id=tokenizer.eos_token_id,
                    pad_token_id=tokenizer.pad_token_id,
                    use_cache=True,
                )
        else:
            outputs = model.generate(
                **inputs,
                max_new_tokens=max_new_tokens,
                do_sample=False,
                repetition_penalty=1.08,
                eos_token_id=tokenizer.eos_token_id,
                pad_token_id=tokenizer.pad_token_id,
                use_cache=True,
            )

    prompt_len = inputs["input_ids"].shape[-1]
    generated = outputs[0][prompt_len:]
    return tokenizer.decode(generated, skip_special_tokens=True).strip()


def evaluate_test_split(
    max_samples: int | None = None,
    report_examples: int = 5,
    max_new_tokens: int = 256,
):
    test_ds = load_dataset(f"{DATASET_WORKSPACE}/{DATASET_REPO}", split="test")
    if "instruction" not in test_ds.column_names or "answer" not in test_ds.column_names:
        raise ValueError("Dataset test split must contain 'instruction' and 'answer' columns.")

    if max_samples is not None:
        max_samples = min(max_samples, len(test_ds))
        test_ds = test_ds.select(range(max_samples))

    n = len(test_ds)
    em_count = 0
    valid_count = 0
    p_sum = 0.0
    r_sum = 0.0
    f1_sum = 0.0
    bad_cases = []

    use_cuda, compute_dtype = _prepare_inference_context()

    start_time = time.perf_counter()
    pbar = tqdm(total=n, desc="Evaluating", unit="sample") if tqdm is not None else None

    for idx, sample in enumerate(test_ds):
        pred = generate_dsl(
            sample["instruction"],
            max_new_tokens=max_new_tokens,
            use_cuda=use_cuda,
            compute_dtype=compute_dtype,
        )
        ref = sample["answer"]

        pred_norm = _normalize_text(pred)
        ref_norm = _normalize_text(ref)

        em = int(pred_norm == ref_norm)
        valid = _is_valid_dsl(pred_norm)
        p, r, f1 = _fact_prf1(_extract_facts(pred_norm), _extract_facts(ref_norm))

        em_count += em
        valid_count += int(valid)
        p_sum += p
        r_sum += r
        f1_sum += f1

        if em == 0 and len(bad_cases) < report_examples:
            bad_cases.append(
                {
                    "idx": idx,
                    "instruction": sample["instruction"][:180],
                    "pred": pred_norm[:300],
                    "ref": ref_norm[:300],
                    "fact_f1": round(f1, 4),
                }
            )

        done = idx + 1
        remaining = n - done
        elapsed = time.perf_counter() - start_time
        avg_per_sample = elapsed / done if done else 0.0
        eta = avg_per_sample * remaining

        if pbar is not None:
            pbar.set_postfix(done=done, remaining=remaining, eta_s=f"{eta:.1f}")
            pbar.update(1)
        elif done % 10 == 0 or done == n:
            print(f"Progress: {done}/{n} | remaining={remaining} | eta~{eta:.1f}s")

    if pbar is not None:
        pbar.close()

    total_time = time.perf_counter() - start_time
    metrics = {
        "num_samples": n,
        "elapsed_seconds": round(total_time, 2),
        "avg_seconds_per_sample": round(total_time / n, 3) if n else 0.0,
        "exact_match": round(em_count / n, 4) if n else 0.0,
        "valid_dsl_rate": round(valid_count / n, 4) if n else 0.0,
        "fact_precision_macro": round(p_sum / n, 4) if n else 0.0,
        "fact_recall_macro": round(r_sum / n, 4) if n else 0.0,
        "fact_f1_macro": round(f1_sum / n, 4) if n else 0.0,
    }

    print("Metrics:")
    for k, v in metrics.items():
        print(f"- {k}: {v}")

    if bad_cases:
        print("\nSample mismatches:")
        for item in bad_cases:
            print(f"\n[idx={item['idx']}] fact_f1={item['fact_f1']}")
            print("instruction:", item["instruction"])
            print("pred:", item["pred"])
            print("ref:", item["ref"])

    return metrics, bad_cases

In [ ]:
metrics_all, examples_all = evaluate_test_split(
    max_samples=None,
    report_examples=10,
    max_new_tokens=256,
)

Metrics:
- num_samples: 338
- elapsed_seconds: 1175.1
- avg_seconds_per_sample: 3.477
- exact_match: 0.0976
- valid_dsl_rate: 0.9822
- fact_precision_macro: 0.6005
- fact_recall_macro: 0.566
- fact_f1_macro: 0.5629

Sample mismatches:

[idx=0] fact_f1=0.25
instruction: Xét tam giác QRS, QT là phân giác
pred: (triangle (Q R S))
(define T point (bisector Q R S))
(on-segment T Q S)
(angle-equal Q R T T S)
ref: (triangle (Q R S))
(define T point (bisector R Q S))
(on-segment T R S)
(segment Q T)

[idx=1] fact_f1=0.6364
instruction: Cho tam giác QRS nhọn, đường cao RV và SW. 1) Chứng minh bốn điểm R, W, V, S cùng thuộc một đường tròn. 2) RV cắt SW tại U; QU cắt RS tại T. Chứng minh RU.RV = RT.RS. 3) Chứng minh
...
(segment C P)
(segment D P)
(segment E P)
(angle-equal P B C P D E)
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...

In [7]:
DATASET_WORKSPACE = "quangne"
DATASET_REPO = "geometry1k"

metrics_all, examples_all = evaluate_test_split(
    max_samples=None,
    report_examples=10,
    max_new_tokens=256,
)

README.md:   0%|          | 0.00/625 [00:00<?, ?B/s]

data/train-00000-of-00001.parquet:   0%|          | 0.00/77.7k [00:00<?, ?B/s]

data/validation-00000-of-00001.parquet:   0%|          | 0.00/15.6k [00:00<?, ?B/s]

data/test-00000-of-00001.parquet:   0%|          | 0.00/16.1k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/749 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/93 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/95 [00:00<?, ? examples/s]

Evaluating:   0%|          | 0/95 [00:00<?, ?sample/s]

Metrics:
- num_samples: 95
- elapsed_seconds: 270.76
- avg_seconds_per_sample: 2.85
- exact_match: 0.1789
- valid_dsl_rate: 0.9895
- fact_precision_macro: 0.8153
- fact_recall_macro: 0.7664
- fact_f1_macro: 0.7738

Sample mismatches:

[idx=1] fact_f1=0.5
instruction: Cho hình bình hành ABCD, điểm P nằm trên đoạn thẳng AB. Tính độ dài đoạn thẳng PD, biết rằng AD = BC.
pred: (parallelogram (A B C D))
(define P point (segment A B))
(segment A D)
(segment B C)
(equal-distance A D B C)
ref: (parallelogram (A B C D))
(segment A B)
(define P point (segment A B))

[idx=3] fact_f1=0.8
instruction: Cho tam giác ABC với góc ABC = 90°, góc BAC = 45°, góc ACB = 45°, đường tròn ngoại tiếp O của tam giác ABC, và điểm N nằm trên đoạn thẳng AB, đường kính AC. Tính bán kính của đường
pred: (triangle (A B C) (right B))
(angle-measure B A C 45)
(angle-measure A C B 45)
(define O point (circumcenter A B C))
(circle O (circumcircle A B C))
(segment A C)
(diameter A C O)
(define N point (segment A B))
(on-ci